### PR TITLE
Handle VIA-pro SMBUS

### PR DESCRIPTION
**Pull request by montjoie**
Handle via-pro SMBUS, permitting to have SPD read. Tested on a a NEC K8T800-8237.

Fixing also an unrelated typo found while read whole document.

Signed-off-by: Corentin Labbe <clabbe.montjoie@gmail.com>

### CODE REVIEW

**Comment by x86fr** [comment]
Can you please post a screenshot here?

**Comment by montjoie** [comment]
You could find everything at http://kernel.montjoie.ovh/zoo/K8T800-8237.html

**Comment by debrouxl** [comment]
I have some computers with a viapro SMBus controller, but I can't test tonight. I'll do my best to test on my side tomorrow evening :)

**Comment by x86fr** [comment]
ok, waiting for @debrouxl. Can you please check if the part number is also missing on your platform?

**Comment by debrouxl** [comment]
When I added the PCI VID+DID lists for unimplemented I2C SMBus-capable controllers, I didn't look further / deeper, so I missed the fact that the viapro and i801 base register maps are the same...

Looking at my `lspci` lists shows that:
* one computer has 1106:3177, i.e. the same device as yours, so we should be able to have a second test tomorrow;
* however, another has 1106:3057, whose SMB base address is 0x90.
The former is one of my usual memtest86+ test benches, the latter isn't and will require some work to become one. Still, it shows that the SMB base address needs to be passed to viapro_get_smb(), instead of hard-coding a 0xD0 SMB base address inside the function.

File: system/smbus.c [hunk 1]
```diff
@@ -1348,6 +1350,24 @@ static bool ich5_get_smb(void)
     return (smbusbase != 0);
 }
 
+// ----------------------
+// VIA SMBUS Controller
+// ----------------------
+// detection based on Linux/i2c-viapro
+static bool viapro_get_smb(void)
```
Comment (debrouxl) [inline system/smbus.c:1357]: SMB base address needs to be passed from the caller, per my comment on the PR :)

**Comment by x86fr** [comment]
It seems that the VIA Pro uses the legacy PIIX4 protocol and registers, so no need to created a duplicate, just link to `piix4_get_smb()`. There is no need for additional redundancy or complexity here.

**Comment by montjoie** [comment]
Same as PIIX except that via Linux driver test for BIT(0) to check if enabled or not.
According to PIIX datasheet https://www.intel.com/Assets/PDF/datasheet/290562.pdf, PIIX should also have this it hardcoded to 1.
But since this function is used for other compatible (amd/serverwork), how to handle this ?

**Comment by debrouxl** [comment]
Indeed, it looks like a variant of this code containing no functional changes, shown below, works well enough on this MSI KT4AV with 1106:3177 :)

P1110518

"Kingston K" is not a typo here, per
```
|00000000 80 08 07 0d 0a 02 40 00-04 50 65 00 82 08 00 01 |??????@ ?Pe ?? ?|
|00000010 0e 04 1c 01 02 20 00 60-70 75 75 38 28 38 28 40 |?????  `puu8(8(@|
|00000020 60 60 40 40 00 00 00 00-00 37 46 20 28 50 00 00 |``@@     7F (P  |
|00000030 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 8c |               ?|
|00000040 7f 98 00 00 00 00 00 00-04 4b 00 00 00 00 00 00 |??      ?K      |
|00000050 00 00 00 00 00 00 00 00-00 00 00 00 00 06 0e 6f |             ??o|
|00000060 08 23 b3 00 00 00 00 00-00 00 00 00 00 00 00 00 |?#?             |
|00000070 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|00000080 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|00000090 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|000000a0 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|000000b0 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|000000c0 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|000000d0 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|000000e0 00 00 00 00 00 00 00 00-00 00 00 00 00 00 00 00 |                |
|000000f0 39 39 30 35 31 39 33 2d-30 31 35 2e 41 30 30 4c |9905193-015.A00L|
```
Smart move putting model data into the vendor-specific area instead of the standard area ?

"Samsung M3 68L6423FTN-CCC (2005-W02)" and "Corsair VS512MB400" are also both correct, including the lack of date information for the latter - both bytes are 0x00 in the raw SPD data.

The patch I used was a variant of the one in this PR, on the path towards unifying `viapro_get_smb()` and `piix4_get_smb()`:
```
diff --git a/system/smbus.c b/system/smbus.c
index d7a2c50..78458e3 100644
--- a/system/smbus.c
+++ b/system/smbus.c
@@ -48,8 +48,9 @@ static uint8_t get_spd(uint8_t slot_idx, uint16_t spd_adr);
 static bool nv_mcp_get_smb(void);
 static bool amd_sb_get_smb(void);
 static bool fch_zen_get_smb(void);
-static bool piix4_get_smb(void);
+static bool piix4_get_smb(int address);
 static bool ich5_get_smb(void);
+static bool viapro_get_smb(int address);
 static uint8_t ich5_process(void);
 static uint8_t ich5_read_spd_byte(uint8_t adr, uint16_t cmd);
 static uint8_t nf_read_spd_byte(uint8_t smbus_adr, uint8_t spd_adr);
@@ -1158,7 +1159,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 return ich5_get_smb();
             }
             if (did == 0x7113) { // 82371AB/EB/MB PIIX4
-                return piix4_get_smb();
+                return piix4_get_smb(0x90);
             }
             // 0x719B 82440/82443MX PMC - PIIX4
             // 0x0F13 ValleyView SMBus Controller ?
@@ -1251,7 +1252,8 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                     // viapro SMBus controller.
                 // case 0x3074: // 8233_0
                 // case 0x3147: // 8233A
-                // case 0x3177: // 8235
+                case 0x3177: // 8235
+                       return viapro_get_smb(0xD0);
                 // case 0x3227: // 8237
                 // case 0x3337: // 8237A
                 // case 0x3372: // 8237S
@@ -1297,7 +1299,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 // case 0x0203: // CSB6
                 // case 0x0205: // HT1000SB
                 // case 0x0408: // HT1100LD
-                    return piix4_get_smb();
+                    return piix4_get_smb(0x90);
                 default:
                     return false;
             }
@@ -1312,9 +1314,9 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
 // PIIX4 SMBUS Controller
 // ----------------------
 
-static bool piix4_get_smb(void)
+static bool piix4_get_smb(int address)
 {
-    uint16_t x = pci_config_read16(0, smbdev, smbfun, 0x90) & 0xFFF0;
+    uint16_t x = pci_config_read16(0, smbdev, smbfun, address) & 0xFFF0;
 
     if (x != 0) {
         smbusbase = x;
@@ -1348,6 +1350,24 @@ static bool ich5_get_smb(void)
     return (smbusbase != 0);
 }
 
+// ----------------------
+// VIA SMBUS Controller
+// ----------------------
+// detection based on Linux/i2c-viapro
+static bool viapro_get_smb(int address)
+{
+    uint16_t x = pci_config_read16(0, smbdev, smbfun, address);
+
+    if (!(x & 1)) {
+        return false;
+    }
+
+    smbusbase = x & 0xFFF0;
+    if (smbusbase == 0)
+        return false;
+    return true;
+}
+
 // --------------------
 // AMD SMBUS Controller
 // --------------------
@@ -1361,7 +1381,7 @@ static bool amd_sb_get_smb(void)
 
     if ((smbus_id & 0xFFFF) == 0x4385 && rev_id <= 0x3D) {
         // Older AMD SouthBridge (SB700 & older) use PIIX4 registers
-        return piix4_get_smb();
+        return piix4_get_smb(0x90);
     } else if ((smbus_id & 0xFFFF) == 0x780B && rev_id == 0x42) {
         // Latest Pre-Zen APUs use the newer Zen PM registers
         return fch_zen_get_smb();
@@ -1568,7 +1588,7 @@ static uint8_t nf_read_spd_byte(uint8_t smbus_adr, uint8_t spd_adr)
     // Start transaction
     __outb(NVSMBCNT_BYTE_DATA | NVSMBCNT_READ, NVSMBCNT);
 
-    // Wait until transction complete
+    // Wait until transaction complete
     for (i = 500; i > 0; i--) {
         usleep(50);
         if (__inb(NVSMBCNT) == 0) {
```

I think that a unified `piix4_get_smb()` should take two arguments: the address, and a boolean for enabling the `if (!(x & 1))` test - if said test is kept at all.

FWIW, the Linux i2c-ali1563.c driver contains:
```
 *	The 1563 southbridge is deceptively similar to the 1533, with a
 *	few notable exceptions. One of those happens to be the fact they
 *	upgraded the i2c core to be 2.0 compliant, and happens to be almost
 *	identical to the i2c controller found in the Intel 801 south
 *	bridges.
 *
 *	This driver is based on a mix of the 15x3, 1535, and i801 drivers,
 *	with a little help from the ALi 1563 spec.
```
and the register map is, accordingly, nearly the same as those of the PIIX4 and i801. I don't have a MB equipped with such a controller, though.

No more tests from me tonight.

**Comment by debrouxl** [comment]
Current state of the patch on my side, after adding and successfully testing support for 1106:3057 (the displayed SPD data matches that displayed by `decode-dimms`, twice in a row; I'll post the picture later):
```
diff --git a/system/smbus.c b/system/smbus.c
index d7a2c50..c0734e6 100644
--- a/system/smbus.c
+++ b/system/smbus.c
@@ -48,8 +48,9 @@ static uint8_t get_spd(uint8_t slot_idx, uint16_t spd_adr);
 static bool nv_mcp_get_smb(void);
 static bool amd_sb_get_smb(void);
 static bool fch_zen_get_smb(void);
-static bool piix4_get_smb(void);
+static bool piix4_get_smb(int address, int check_enable);
 static bool ich5_get_smb(void);
+//static bool viapro_get_smb(int address);
 static uint8_t ich5_process(void);
 static uint8_t ich5_read_spd_byte(uint8_t adr, uint16_t cmd);
 static uint8_t nf_read_spd_byte(uint8_t smbus_adr, uint8_t spd_adr);
@@ -1158,7 +1159,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 return ich5_get_smb();
             }
             if (did == 0x7113) { // 82371AB/EB/MB PIIX4
-                return piix4_get_smb();
+                return piix4_get_smb(0x90, false);
             }
             // 0x719B 82440/82443MX PMC - PIIX4
             // 0x0F13 ValleyView SMBus Controller ?
@@ -1243,15 +1244,16 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                     // via SMBus controller.
                 // case 0x3050: // 82C596_3
                     // Try SMB base address = 0x90, then SMB base address = 0x80
-                    // viapro SMBus controller.
+                    // viapro SMBus controller, i.e. PIIX4 with a small quirk.
                 // case 0x3051: // 82C596B_3
-                // case 0x3057: // 82C686_4
+                case 0x3057: // 82C686_4
                 // case 0x8235: // 8231_4
                     // SMB base address = 0x90
-                    // viapro SMBus controller.
+                    // viapro SMBus controller, i.e. PIIX4 with a small quirk.
+                    return piix4_get_smb(0x90, true);
                 // case 0x3074: // 8233_0
                 // case 0x3147: // 8233A
-                // case 0x3177: // 8235
+                case 0x3177: // 8235
                 // case 0x3227: // 8237
                 // case 0x3337: // 8237A
                 // case 0x3372: // 8237S
@@ -1261,7 +1263,8 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 // case 0x8409: // VX855
                 // case 0x8410: // VX900
                     // SMB base address = 0xD0
-                    // viapro I2C controller.
+                    // viapro SMBus controller, i.e. PIIX4 with a small quirk.
+                    return piix4_get_smb(0xD0, true);
                 default:
                     return false;
             }
@@ -1279,7 +1282,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
         case PCI_VID_ALI:
             switch(did)
             {
-                // case 0x1563: // ali1563 (M1563) SMBus controller
+                // case 0x1563: // ali1563 (M1563) SMBus controller, nearly compatible with PIIX4 according to Linux i2c-ali1563 driver.
                 // case 0x7101: // ali1535 (M1535) or ali15x3 (M1533/M1543) SMBus controllers
                 default:
                     return false;
@@ -1297,7 +1300,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 // case 0x0203: // CSB6
                 // case 0x0205: // HT1000SB
                 // case 0x0408: // HT1100LD
-                    return piix4_get_smb();
+                    return piix4_get_smb(0x90, false);
                 default:
                     return false;
             }
@@ -1312,10 +1315,15 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
 // PIIX4 SMBUS Controller
 // ----------------------
 
-static bool piix4_get_smb(void)
+static bool piix4_get_smb(int address, int check_enable)
 {
-    uint16_t x = pci_config_read16(0, smbdev, smbfun, 0x90) & 0xFFF0;
+    uint16_t x = pci_config_read16(0, smbdev, smbfun, address);
 
+    if (check_enable && !(x & 1)) {
+        return false;
+    }
+
+    x &= 0xFFF0;
     if (x != 0) {
         smbusbase = x;
         return true;
@@ -1348,6 +1356,24 @@ static bool ich5_get_smb(void)
     return (smbusbase != 0);
 }
 
+// ----------------------
+// VIA SMBUS Controller
+// ----------------------
+// detection based on Linux/i2c-viapro
+/*static bool viapro_get_smb(int address)
+{
+    uint16_t x = pci_config_read16(0, smbdev, smbfun, address);
+
+    if (!(x & 1)) {
+        return false;
+    }
+
+    smbusbase = x & 0xFFF0;
+    if (smbusbase == 0)
+        return false;
+    return true;
+}*/
+
 // --------------------
 // AMD SMBUS Controller
 // --------------------
@@ -1361,7 +1387,7 @@ static bool amd_sb_get_smb(void)
 
     if ((smbus_id & 0xFFFF) == 0x4385 && rev_id <= 0x3D) {
         // Older AMD SouthBridge (SB700 & older) use PIIX4 registers
-        return piix4_get_smb();
+        return piix4_get_smb(0x90, false);
     } else if ((smbus_id & 0xFFFF) == 0x780B && rev_id == 0x42) {
         // Latest Pre-Zen APUs use the newer Zen PM registers
         return fch_zen_get_smb();
@@ -1568,7 +1594,7 @@ static uint8_t nf_read_spd_byte(uint8_t smbus_adr, uint8_t spd_adr)
     // Start transaction
     __outb(NVSMBCNT_BYTE_DATA | NVSMBCNT_READ, NVSMBCNT);
 
-    // Wait until transction complete
+    // Wait until transaction complete
     for (i = 500; i > 0; i--) {
         usleep(50);
         if (__inb(NVSMBCNT) == 0) {
```

Regression testing:
- [x] 1002:4385 (AMD PIIX4) in ASUS RS904A with 4 x Opteron 6272: SPD data successfully retrieved;
- [ ] 8086:7113 (Intel PIIX4) in HP Kayak XW with 2 x Xeon PII 350 QS;
- [ ] 1166:0201 (Serverworks CSB5) in IBM xSeries 345 on 2 x Xeon P4 S604.

**Comment by debrouxl** [comment]
Unsurprisingly, Intel PIIX4, Serverworks CSB5, and viapro did not regress with the following patch, which looks very much like a final version:
```
diff --git a/system/smbus.c b/system/smbus.c
index d7a2c50..463e584 100644
--- a/system/smbus.c
+++ b/system/smbus.c
@@ -48,7 +48,7 @@ static uint8_t get_spd(uint8_t slot_idx, uint16_t spd_adr);
 static bool nv_mcp_get_smb(void);
 static bool amd_sb_get_smb(void);
 static bool fch_zen_get_smb(void);
-static bool piix4_get_smb(void);
+static bool piix4_get_smb(uint8_t address);
 static bool ich5_get_smb(void);
 static uint8_t ich5_process(void);
 static uint8_t ich5_read_spd_byte(uint8_t adr, uint16_t cmd);
@@ -1158,7 +1158,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 return ich5_get_smb();
             }
             if (did == 0x7113) { // 82371AB/EB/MB PIIX4
-                return piix4_get_smb();
+                return piix4_get_smb(0x90);
             }
             // 0x719B 82440/82443MX PMC - PIIX4
             // 0x0F13 ValleyView SMBus Controller ?
@@ -1243,15 +1243,16 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                     // via SMBus controller.
                 // case 0x3050: // 82C596_3
                     // Try SMB base address = 0x90, then SMB base address = 0x80
-                    // viapro SMBus controller.
+                    // viapro SMBus controller, i.e. PIIX4 with a small quirk.
                 // case 0x3051: // 82C596B_3
-                // case 0x3057: // 82C686_4
+                case 0x3057: // 82C686_4
                 // case 0x8235: // 8231_4
                     // SMB base address = 0x90
-                    // viapro SMBus controller.
+                    // viapro SMBus controller, i.e. PIIX4.
+                    return piix4_get_smb(0x90);
                 // case 0x3074: // 8233_0
                 // case 0x3147: // 8233A
-                // case 0x3177: // 8235
+                case 0x3177: // 8235
                 // case 0x3227: // 8237
                 // case 0x3337: // 8237A
                 // case 0x3372: // 8237S
@@ -1261,7 +1262,8 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 // case 0x8409: // VX855
                 // case 0x8410: // VX900
                     // SMB base address = 0xD0
-                    // viapro I2C controller.
+                    // viapro SMBus controller, i.e. PIIX4 with a small quirk.
+                    return piix4_get_smb(0xD0);
                 default:
                     return false;
             }
@@ -1279,7 +1281,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
         case PCI_VID_ALI:
             switch(did)
             {
-                // case 0x1563: // ali1563 (M1563) SMBus controller
+                // case 0x1563: // ali1563 (M1563) SMBus controller, nearly compatible with PIIX4 according to Linux i2c-ali1563 driver.
                 // case 0x7101: // ali1535 (M1535) or ali15x3 (M1533/M1543) SMBus controllers
                 default:
                     return false;
@@ -1297,7 +1299,7 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
                 // case 0x0203: // CSB6
                 // case 0x0205: // HT1000SB
                 // case 0x0408: // HT1100LD
-                    return piix4_get_smb();
+                    return piix4_get_smb(0x90);
                 default:
                     return false;
             }
@@ -1312,9 +1314,9 @@ static bool find_smb_controller(uint16_t vid, uint16_t did)
 // PIIX4 SMBUS Controller
 // ----------------------
 
-static bool piix4_get_smb(void)
+static bool piix4_get_smb(uint8_t address)
 {
-    uint16_t x = pci_config_read16(0, smbdev, smbfun, 0x90) & 0xFFF0;
+    uint16_t x = pci_config_read16(0, smbdev, smbfun, address) & 0xFFF0;
 
     if (x != 0) {
         smbusbase = x;
@@ -1361,7 +1363,7 @@ static bool amd_sb_get_smb(void)
 
     if ((smbus_id & 0xFFFF) == 0x4385 && rev_id <= 0x3D) {
         // Older AMD SouthBridge (SB700 & older) use PIIX4 registers
-        return piix4_get_smb();
+        return piix4_get_smb(0x90);
     } else if ((smbus_id & 0xFFFF) == 0x780B && rev_id == 0x42) {
         // Latest Pre-Zen APUs use the newer Zen PM registers
         return fch_zen_get_smb();
@@ -1568,7 +1570,7 @@ static uint8_t nf_read_spd_byte(uint8_t smbus_adr, uint8_t spd_adr)
     // Start transaction
     __outb(NVSMBCNT_BYTE_DATA | NVSMBCNT_READ, NVSMBCNT);
 
-    // Wait until transction complete
+    // Wait until transaction complete
     for (i = 500; i > 0; i--) {
         usleep(50);
         if (__inb(NVSMBCNT) == 0) {
```

Author: Corentin Labbe <clabbe.montjoie@gmail.com>
Co-developed-by: Lionel Debroux <lionel_debroux@yahoo.fr>
Co-developed-by: Sam Demeulemeester <github@x86-secret.com>

**Comment by x86fr** [comment]
Looks good (pity that C doesn't support default values for function variables...)

**Comment by x86fr** [comment]
I'm about to commit the patch, but I can't see any screenshot for VIA:0x3057. 
Is it working on both 686A and 686B ?

**Comment by x86fr** [comment]
Pushed in commit 10e8435

**Comment by x86fr** [comment]
Cool! The patch has been included in the commit, so everything's OK. :)

**Pull request closed without merging**